### PR TITLE
[Kogito-7981] Provide a Build CRD for the kaniko Builder

### PR DESCRIPTION
**Pull request by desmax74**
See: https://issues.redhat.com/browse/KOGITO-7981
Signed-off-by: desmax74 <mdessi@redhat.com>

**Description of the change:**

**Motivation for the change:**

**Checklist**

- [ ] Add or Modify a unit test for your change
- [ ] Have you verified that tall the tests are passing?

### CODE REVIEW

File: api/v08/kogitoserverless_build_types.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+	"github.com/RHsyseng/operator-utils/pkg/olm"
+	"github.com/ricardozanini/kogito-builder/api"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// KogitoServerlessBuildSpec defines the desired state of KogitoServerlessBuild
+type KogitoServerlessBuildSpec struct {
```
Comment (spolti) [inline api/v08/kogitoserverless_build_types.go:29]: please apply the kubebuilder annotations so the types are correctly added to the crd.
please add to all types.

File: api/v08/kogitoserverless_build_types.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+	SwfName       string `json:"SwfName,omitempty"`
```
Comment (ricardozanini) [inline api/v08/kogitoserverless_build_types.go:32]: I think we should avoid abbreviations. What about `WorkflowName`?

File: api/v08/kogitoserverless_build_types.go [hunk 3]
```diff
@@ -25,31 +25,41 @@ import (
...
 // KogitoServerlessBuildSpec defines the desired state of KogitoServerlessBuild
 type KogitoServerlessBuildSpec struct {
 	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
 	// Important: Run "make" to regenerate code after modifying this file
-	SwfName       string `json:"SwfName,omitempty"`
-	ImageName     string `json:"ImageName,omitempty"`
-	PodMiddleName string `json:"PodMiddleName,omitempty"`
-	SourceSwf     []byte
-	Dockerfile    []byte
+	SwfName         string `json:"SwfName,omitempty"`
```
Comment (spolti) [inline api/v08/kogitoserverless_build_types.go:35]: please add comments on each field to enrich the crd.

File: api/v08/kogitoserverless_build_types.go [hunk 4]
```diff
@@ -25,31 +25,41 @@ import (
...
 // KogitoServerlessBuildStatus defines the observed state of KogitoServerlessBuild
 type KogitoServerlessBuildStatus struct {
 	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
 	// Important: Run "make" to regenerate code after modifying this file
-	Deployments olm.DeploymentStatus      `json:"deployments"`
-	Phase       ConditionType             `json:"phase,omitempty"`
-	Applied     KogitoServerlessBuildSpec `json:"applied,omitempty"`
-	Version     string                    `json:"version,omitempty"`
-	BuildPhase  api.BuildPhase            `json:"BuildPhase,omitempty"`
+	SwfName         string                    `json:"SwfName,omitempty"`
```
Comment (spolti) [inline api/v08/kogitoserverless_build_types.go:50]: same

File: controllers/utils.go [hunk 5]
```diff
@@ -51,41 +53,46 @@ func GeneratePassword(length int) []byte {
...
-func GetBuild(ctx context.Context, req ctrl.Request, workflowID string, r *KogitoServerlessWorkflowReconciler) (apiv08.KogitoServerlessBuild, error) {
+func GetBuild(ctx context.Context, req ctrl.Request, workflowID string, r *KogitoServerlessWorkflowReconciler, log logr.Logger) (apiv08.KogitoServerlessBuild, error) {
 	buildInstance := &apiv08.KogitoServerlessBuild{}
 	buildInstance.Spec.SwfName = workflowID
 	error := r.Client.Get(ctx, types.NamespacedName{Namespace: req.NamespacedName.Name, Name: workflowID}, buildInstance)
 	return *buildInstance, error
 }
 
-func CreateBuild(ctx context.Context, workflowID string, workflowSwf []byte, r *KogitoServerlessWorkflowReconciler) (apiv08.KogitoServerlessBuild, error) {
+func CreateBuild(ctx context.Context, workflowID string, workflowSwf []byte, r *KogitoServerlessWorkflowReconciler, log logr.Logger) (apiv08.KogitoServerlessBuild, error) {
```
Comment (ricardozanini) [inline controllers/utils.go:63]: Why are you adding the logger here if it's not being used? I liked the Camel-K approach we replicated in the builder, where we implemented `InjectLogger` and delegated it to the caller to inject it. This way, we avoid passing too many arguments. What about considering this approach here?

File: controllers/kogitoserverlessworkflow_controller.go [hunk 6]
```diff
@@ -78,22 +86,22 @@ func (r *KogitoServerlessWorkflowReconciler) Reconcile(ctx context.Context, req
...
-	_, err = builder.BuildImageWithDefaults(workflow.ID, jsonWorkflow)
-	if err != nil {
-		log.Info("Error building KogitoServerlessWorkflow into Workflow: ", "message", err.Error())
-	}
-	return ctrl.Result{}, nil
+	buildable := builder.NewBuildable(r.Client, ctx)
+	build, err := buildable.HandleSwfBuild(workflow.ID, jsonWorkflow, req)
+	log.Info(string(build.Spec.BuildPhase))
+	log.Info(string(build.Status.BuildPhase))
+	return ctrl.Result{Requeue: true}, err
```
Comment (davidesalerno) [inline controllers/kogitoserverlessworkflow_controller.go:93]: Why are we requeue the Workflow CR?

File: builder/builder_pod.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+	return Buildable{
+		Ctx:    Ctx,
+		client: client,
+	}
+}
+
+func (buildable *Buildable) GetSwfBuild(ctx context.Context, req ctrl.Request, workflowID string, client client.Client) (apiv08.KogitoServerlessBuild, error) {
+	buildInstance := &apiv08.KogitoServerlessBuild{}
+	buildInstance.Spec.SwfName = workflowID
+	error := client.Get(ctx, types.NamespacedName{Namespace: req.NamespacedName.Name, Name: workflowID}, buildInstance)
```
Comment (davidesalerno) [inline builder/builder_pod.go:44]: Here you are checking if there is a Build CR into the req.NamespacedName.Name namespace and at line 52 you are creating it into another namespace (BUILDER_NAMESPACE_DEFAULT).
 Why?

File: controllers/kogitoserverlessbuild_controller.go [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+	instance := &api08.KogitoServerlessBuild{}
+
+	err := r.Get(ctx, req.NamespacedName, instance)
+	if err == nil {
+		log.Info("Build CR present")
+		phase := instance.Spec.BuildPhase
+		log.Info(string(phase))
+		if phase != api.BuildPhaseSucceeded && phase != api.BuildPhaseError && phase != api.BuildPhaseFailed && phase != api.BuildPhasePending {
+			builder := builder.NewBuilder(ctx)
+			build, err := builder.BuildImageWithDefaults(instance.Spec.SwfName, instance.Spec.SourceSwf)
```
Comment (davidesalerno) [inline controllers/kogitoserverlessbuild_controller.go:67]: Since the BuildImageWithDefaults is calling the kogito-builder Schedule function, every time you are calling this, you will schedule a pod

File: Makefile [hunk 9]
```diff
@@ -29,7 +29,8 @@ BUNDLE_METADATA_OPTS ?= $(BUNDLE_CHANNELS) $(BUNDLE_DEFAULT_CHANNEL)
 #
 # For example, running 'make bundle-build bundle-push catalog-build catalog-push' will build and push both
 # kiegroup.org/kogito-serverless-operator-bundle:$VERSION and kiegroup.org/kogito-serverless-operator-catalog:$VERSION.
-IMAGE_TAG_BASE ?= quay.io/dsalerno/kogito-serverless-operator
+IMAGE_TAG_BASE ?= quay.io/mdessi/kogito-serverless-operator
```
Comment (davidesalerno) [inline Makefile:32]: I think that it is better to revert this change so that when we will move the project to kiegroup org the replace will be easier.

File: Makefile [hunk 10]
```diff
@@ -39,7 +40,8 @@ BUNDLE_IMG ?= $(IMAGE_TAG_BASE)-bundle:v$(VERSION)
 BUNDLE_GEN_FLAGS ?= -q --overwrite --version $(VERSION) $(BUNDLE_METADATA_OPTS)
 
 # Container runtime engine used for building the images
-BUILDER ?= podman
+BUILDER ?= docker
```
Comment (davidesalerno) [inline Makefile:43]: Since we switched from docker to podman as a default container manager on all the other projects it is better to use podman by default.

If you need to use docker you can simply use the BUILDER variable without changing the Makefile

File: Makefile (hunk 9 above)
Comment (desmax74) [reply Makefile:32]: ok

File: Makefile (hunk 10 above)
Comment (desmax74) [reply Makefile:43]: ok

File: Dockerfile [hunk 11]
```diff
@@ -16,13 +16,15 @@ COPY controllers/ controllers/
...
+COPY utils/ utils/
 
 # Build
 RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -a -o manager main.go
 
 # Use distroless as minimal base image to package the manager binary
 # Refer to https://github.com/GoogleContainerTools/distroless for more details
-FROM gcr.io/distroless/static:nonroot
+#FROM gcr.io/distroless/static:nonroot
+FROM registry.access.redhat.com/ubi8/ubi-minimal:latest
```
Comment (spolti) [inline Dockerfile:27]: there is now the ubi-micro, maybe its worth to try to use that and see how it goes :)

File: Makefile [hunk 12]
```diff
@@ -115,10 +115,18 @@ test: manifests generate fmt vet envtest ## Run tests.
...
+build-4-debug: generate fmt vet ## Build manager binary with debug options.
+	go build -gcflags="all=-N -l" -o bin/manager main.go
+
 .PHONY: run
 run: manifests generate fmt vet ## Run a controller from your host.
 	go run ./main.go
 
+.PHONY: debug
+debug: build-4-debug ## Run a controller from your host from binary
+	./bin/manager
```
Comment (spolti) [inline Makefile:128]: what this debug here is doing?

File: Makefile [hunk 13]
```diff
@@ -185,23 +193,23 @@ ENVTEST ?= $(LOCALBIN)/setup-envtest
...
 KUSTOMIZE_VERSION ?= v3.8.7
-CONTROLLER_TOOLS_VERSION ?= v0.8.0
+CONTROLLER_TOOLS_VERSION ?= v0.9.2
 
 KUSTOMIZE_INSTALL_SCRIPT ?= "https://raw.githubusercontent.com/kubernetes-sigs/kustomize/master/hack/install_kustomize.sh"
 .PHONY: kustomize
 kustomize: $(KUSTOMIZE) ## Download kustomize locally if necessary.
 $(KUSTOMIZE): $(LOCALBIN)
-	curl -s $(KUSTOMIZE_INSTALL_SCRIPT) | bash -s -- $(subst v,,$(KUSTOMIZE_VERSION)) $(LOCALBIN)
+	test -s $(LOCALBIN)/kustomize || { curl -s $(KUSTOMIZE_INSTALL_SCRIPT) | bash -s -- $(subst v,,$(KUSTOMIZE_VERSION)) $(LOCALBIN); }
```
Comment (spolti) [inline Makefile:202]: make sure to align this changes with the operator-sdk suggestions for the specific version you are using.

File: README.md [hunk 14]
```diff
@@ -100,6 +101,18 @@ kubectl create namespace greeting-workflow
 kubectl apply -f config/samples/sw.kogito.kie.org__v08_kogitoserverlessworkflow.yaml -n greeting-workflow
 ```
 
+8. Default configuration
+
+A configmap called kogito-builder-config will be created under the kogito-builder namespace at the first application of a workflow, it contains:
```
Comment (spolti) [inline README.md:106]: can we add a example here?

File: Makefile (hunk 12 above)
Comment (desmax74) [reply Makefile:128]: One run the build with the flags for debug this run the debug after the build

File: Makefile (hunk 13 above)
Comment (desmax74) [reply Makefile:202]: Copied from 1.23.0 version of the operator https://sdk.operatorframework.io/docs/upgrading-sdk-version/v1.23.0/ during the migration from 1.22.0 to 1.24.0

File: README.md [hunk 15]
```diff
@@ -100,6 +101,18 @@ kubectl create namespace greeting-workflow
...
+  
+- BUILDER_RESOURCE_DEFAULT = Dockerfile
+- WORKFLOW_DEFAULT_EXTENSION = .sw.json
+- DEFAULT_KANIKO_SECRET_DEFAULT = regcred
+- DEFAULT_REGISTRY_REPO =  quay.io/kiegroup
+- Dockerfile = dockerfile content
+
+For the local development the DEFAULT_REGISTRY_REPO must be changed
+
 ## Contributing
```
Comment (spolti) [inline README.md:116]: would be nice to break this in another .md file like we can find on another projects.

File: Makefile [hunk 16]
```diff
@@ -185,23 +193,23 @@ ENVTEST ?= $(LOCALBIN)/setup-envtest
...
 
 .PHONY: controller-gen
 controller-gen: $(CONTROLLER_GEN) ## Download controller-gen locally if necessary.
 $(CONTROLLER_GEN): $(LOCALBIN)
-	GOBIN=$(LOCALBIN) go install sigs.k8s.io/controller-tools/cmd/controller-gen@$(CONTROLLER_TOOLS_VERSION)
+	test -s $(LOCALBIN)/controller-gen || GOBIN=$(LOCALBIN) go install sigs.k8s.io/controller-tools/cmd/controller-gen@$(CONTROLLER_TOOLS_VERSION)
```
[the remark this answers is not in the document]
Comment (desmax74) [reply Makefile:207]: Basically it checks if the binaries are already present instead to download every time https://sdk.operatorframework.io/docs/upgrading-sdk-version/v1.23.0/

File: README.md (hunk 14 above)
Comment (desmax74) [reply README.md:106]: A screenshot or the yaml ?

File: Dockerfile (hunk 11 above)
Comment (desmax74) [reply Dockerfile:27]: @spolti with minimal the operator image is 150 mb with micro the operator image 82.8 mb , switching to micro

**Review by spolti — CHANGES_REQUESTED** [review]

**Comment by spolti** [comment]
Nice work guys :D

added just a few comments, nothing critical :)

File: api/v08/kogitoserverlessbuild_types.go [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+// KogitoServerlessBuildSpec defines the desired state of KogitoServerlessBuild
+type KogitoServerlessBuildSpec struct {
+	//Workflow's unique identifier
+	WorkflowId string `json:"workflowId,omitempty"`
+	//Image name
+	ImageName string `json:"imageName,omitempty"`
+	// Middlename of the pod
+	PodMiddleName string `json:"podMiddleName,omitempty"`
+	//Dockerfile content used for the build @TODO load from configmap
+	Dockerfile []byte `json:"dockerfile,omitempty"`
```
[the remark this answers is not in the document]
Comment (desmax74) [reply api/v08/kogitoserverlessbuild_types.go:33]: ok

File: builder/builder.go [hunk 18]
```diff
@@ -17,89 +17,132 @@ package builder
...
+func NewBuilder(contex context.Context, cm corev1.ConfigMap) Builder {
+	return Builder{ctx: contex, cm: cm}
 }
 
-func (b *Builder) BuildImageWithDefaults(sourceSwfName string, sourceSwf []byte) (*api.Build, error) {
-	wd, _ := os.Getwd()
-	dockerFile, _ := os.ReadFile(wd + "/builder/Dockerfile")
-	ib := NewImageBuilder(sourceSwfName, sourceSwf, dockerFile)
+func (b *Builder) getImageBuilder(workflowID string, workflowDefinition []byte) ImageBuilder {
+	dockerfile := b.cm.Data[constants.BUILDER_RESOURCE_NAME_DEFAULT]
```
[the remark this answers is not in the document]
Comment (desmax74) [reply builder/builder.go:40]: ok

File: builder/builder_test.go [hunk 19]
```diff
@@ -1,18 +1,13 @@
...
-	"os"
 	"testing"
-	"time"
 )
 
 // Minikube must be up and running with the default BUILDER_NAMESPACE_DEFAULT available,
 // if the repo is not available on quay, will be created with the first pull but wil be created private
 func TestBuild(t *testing.T) {
-	wd, _ := os.Getwd()
+	/*wd, _ := os.Getwd()
```
[the remark this answers is not in the document]
Comment (desmax74) [reply builder/builder_test.go:10]: ok

File: builder/Dockerfile [hunk 20]
```diff
@@ -20,10 +13,10 @@ FROM registry.access.redhat.com/ubi8/openjdk-11:latest
 ENV LANG='en_US.UTF-8' LANGUAGE='en_US:en'
 
 # We make four distinct layers so if there are application changes the library layers can be re-used
-COPY --from=builder --chown=185 /home/kogito/kogito-base/target/quarkus-app/lib/ /deployments/lib/
-COPY --from=builder --chown=185 /home/kogito/kogito-base/target/quarkus-app/*.jar /deployments/
-COPY --from=builder --chown=185 /home/kogito/kogito-base/target/quarkus-app/app/ /deployments/app/
-COPY --from=builder --chown=185 /home/kogito/kogito-base/target/quarkus-app/quarkus/ /deployments/quarkus/
+COPY --from=builder --chown=185 /home/kogito/serverless-workflow-project/target/quarkus-app/lib/ /deployments/lib/
```
[the remark this answers is not in the document]
Comment (desmax74) [reply builder/Dockerfile:16]: Ok, moved all under /home/kogito/bin/

File: README.md (hunk 15 above)
Comment (desmax74) [reply README.md:116]: When will be merged with the platform

File: README.md (hunk 14 above)
Comment (desmax74) [reply README.md:106]: Or better we provide an install section olm/script
